File: lib/test/structuredfield/decodeSfList.test.ts
import assert from 'node:assert';
import { describe, it } from 'node:test';
import { SfItem } from '../../src/structuredfield/SfItem.js';
import { decodeSfList } from '../../src/structuredfield/decodeSfList.js';

describe('decodeSfList', () => {
	it('handles an empty string', () => {
		assert.deepStrictEqual(decodeSfList(``), []);
	});

	it('parses valid input', () => {
		assert.deepStrictEqual(decodeSfList(`("foo"; a=1;b=2);lvl=5, ("bar" "baz");lvl=1`), [
			new SfItem([
				new SfItem('foo', { 'a': 1, 'b': 2 }),
			], { 'lvl': 5 }),
			new SfItem(['bar', 'baz'], { 'lvl': 1 }),
		]);
	});

	it('throws an error on invalid input', () => {
		assert.throws(() => decodeSfList(`1,2,3)`), (error: any) => {
			assert.deepStrictEqual(error.message, `failed to parse "1,2,3)" as List`);
			assert.deepStrictEqual(error.cause.message, `failed to parse ")" as List`);
			return true;
		});
		assert.throws(() => decodeSfList(`1,2,`), (error: any) => {
			assert.deepStrictEqual(error.message, `failed to parse "1,2," as List`);
			assert.deepStrictEqual(error.cause.message, `failed to parse "" as List`);
			return true;
		});
	});
});

File: python/src/test/structuredfield/decodeSfList_test.py
import unittest

from ...structuredfield.decodeSfList import decodeSfList
from ...structuredfield.SfException import SfException
from ...structuredfield.SfItem import SfItem
from ...structuredfield.SfToken import SfToken


class TestDecodeSfItem(unittest.TestCase):
    def test_empty_string(self):
        self.assertEqual(decodeSfList(""), [])

    def test_valid_input(self):
        self.assertEqual(
            decodeSfList('("foo"; a=1;b=2);lvl=5, ("bar" "baz");lvl=1'),
            [
                SfItem(
                    [
                        SfItem("foo", {"a": 1, "b": 2}),
                    ],
                    {"lvl": 5},
                ),
                SfItem(["bar", "baz"], {"lvl": 1}),
            ],
        )

    def test_invalid_input(self):
        with self.assertRaises(SfException) as context:
            decodeSfList("1,2,3)")
        self.assertEqual(context.exception.message, 'failed to parse "1,2,3)" as List')
        self.assertEqual(context.exception.cause.message, 'failed to parse ")" as List')

        with self.assertRaises(SfException) as context:
            decodeSfList("1,2,")
        self.assertEqual(context.exception.message, 'failed to parse "1,2," as List')
        self.assertEqual(context.exception.cause.message, 'failed to parse "" as List')
